### ISSUE TITLE
How to keep this repo in sync with the original development line?

### ISSUE THREAD

**Issue by Foadsf** [SimulIDE/SimulIDE#9]
Sadly the original developer of the software does not comply with the common programming best practices. He used to publish binaries and zip files on SourceForge, then he switched to his blog. and right now he only publishes source code as zip files plus binaries (only for Windows?) on his Patreon. I have tried to collect whatever code I could find on the internet and then convert the svn structure, to git. we now have several repositories:

- [This repo](https://github.com/SimulIDE/SimulIDE), which I hope to become the main line of development in the future
- The SimulIDE03 repo, which is what I have been working on trying to make the [latest version at the time](https://github.com/SimulIDE/SimulIDE_0.3.12-SR8_Sources) of the fork to a stable cross-platform software.
- there are the [SimulIDE_0.3.12-SR8_Sources and ](https://github.com/SimulIDE/SimulIDE_0.3.12-SR8_Sources) and [SimulIDE_0.4.13-Final_Sources](https://github.com/SimulIDE/SimulIDE_0.4.13-Final_Sources), which are the duplicates of what the original developer published on his Pateron.
- The [SimulIDE_SF](https://github.com/SimulIDE/SimulIDE_SF) which is a duplicate of what I could find on SourceForge.

As you may see it is a mess. I'm looking forward to seeing your ideas that how we can clean up things and merge everything in or two repositories.

**Comment by ghost** [comment]
IMHO, there are two different tasks:

First, merge all the projects, and try to keep the history (SimulIDE_SF seems to have a lot of commits, I would keep them), and use tags or release branches as a starting point.

Once you have your standard branches clearly defined (i.e. develop, release/0.3.12, etc.), use a "vendor branch" for the code that comes from Patreon.

**Comment by pratikpc** [comment]
Also are there benefits of keeping https://github.com/SimulIDE/SimulIDE/tree/master/src/simavr up-to-date with https://github.com/buserror/simavr?

**Comment by Mihaylov93** [comment]
There should be but there also could bring unexpected issues, I would say instead of having it inside the repo (Usually bad practice) we should use git submodules.

**Comment by pratikpc** [comment]
@Mihaylov93 agree with you

But Git Submodules do not support adding only a subdirectory right?

**Comment by Mihaylov93** [comment]
Yes, there is ways to set the subfolder as a submodule but I agree with this answer here https://stackoverflow.com/a/5303850
I would live with having the whole repo as submodule

Another possibility is to have a fork and use it in this project, that way we also have some form of versioning.

**Comment by Minetrain2002** [comment]
what about this https://launchpad.net/simulide
was linked to on the blog here https://www.simulide.com/p/home.html

**Comment by ppenguin** [comment]
Using [this](https://stackoverflow.com/a/61983069/12771809) and [this](https://bazaar.launchpad.net/~arcachofo/simulide/1.0.1/files) I could download the most recent source tree with one `curl` command:

```
curl -SLo simulide.tgz "https://bazaar.launchpad.net/~arcachofo/simulide/1.0.1/tarball/1425"
```

I guess it should in theory be possible to make a `github action` that does this automatically (possibly would need timed execution and some scraping to find the most recent revision), and then automatically merge the changes into the chosen branch of this repo, along the lines of (executed e.g. every week via GHA cron functionality)

1. create and checkout in this repo the branch `upstream-copy`
2. download and unpack the source tree according to the above, e.g. sth like
    ```
    curl -sL "https://bazaar.launchpad.net/~arcachofo/simulide/1.0.1/tarball/1425" | tar -zxv -C $(pwd) --strip-components=3 -f -
    ```
4. commit and push
5. checkout our tracking branch, e.g. `upstream`
6. merge `upstream-copy` into `upstream`
7. delete `upstream-copy`

Of course if one is willing to occasionally perform these steps manually (the number of commits seems to be between 0 and about 10 / month) these steps aren't so time consuming...

One could then have an own `main` which often merges changes from `upstream` but contains additional improvements...

(I couldn't make [these instructions](https://help.launchpad.net/Code/Git) work for me to directly clone the repo with `git`, might also be related to restrictions that are on the repo only download with password?)

**Comment by ppenguin** [comment]
As a test and PoC I made a [`nix flake`](https://github.com/ppenguin/SimulIDE/tree/from-scratch-upstream-trunk-1723)  that can build and run _almost directly_ (it needed some minor fixes) from the current upstream trunk, can be tested like so:

```
nix run "github:/ppenguin/SimulIDE/?ref=from-scratch-upstream-trunk-1723&dir=.nix#simulide"
```

This is mainly to prove that upstream can be built declaratively, which means virtually no maintenance overhead.
It may as well be possible to use `brz` in the `nix` derivation, thereby making it act on a "live-pulled" revision from `launchpad`. While it would arguably be "better" if upstream would use `git`, this is (for `nix` users at least) a feasible way do do spin-offs that have relatively minor modifications (any that can be added by patching).

**Comment by albfan** [comment]
Probably all this can be rebuild, now original project has a git repo:

release: https://git.launchpad.net/simulide
develop: https://github.com/eeTools/SimulIDE-dev 

Use submodule if you use another git repo is a must

Probably all these is related with #32
